urlc98c3b97d2_Trimax_venta_tgz-pJ8-com_gesoftware_venta_structures_map_MapTestJ8/src/com/gesoftware/venta/network/Server.java

```diff
@@ -0,0 +1,570 @@
+                            m_IsClientActive = false;
+                    } catch (final Exception e) {
+                        LoggingUtility.debug("Exception: " + e.getMessage());
+                        LoggingUtility.core("Client closed connection: " + m_ID);
+                        m_IsClientActive = false;
+                    }
+                }
+            } /* End of 'Server::loop' method */
+
+            @Override
+            public final void run() {
+                LoggingUtility.core("Receiver thread started for client: " + m_ID);
+
+                try {
+                    loop();
+                } catch (final Exception e) {
+                    LoggingUtility.debug("Error during client loop: " + e.getMessage());
+                }
+
+                /* Closing input stream */
+                try {
+                    m_Reader.close();
+                } catch (final Exception e) {
+                    LoggingUtility.error("Can't close input stream: " + e.getMessage());
+                }
+
+                /* Disconnect client from server */
+                disconnectClient();
+
+                /* Execute onDisconnect method of handler */
+                m_ServerHandler.onDisconnect(m_ID);
+                LoggingUtility.core("Receiver thread closed for client: " + m_ID);
+            } /* End of 'ClientReceiver::run' method */
+
+            /* *
+             * METHOD: Stops receiver immediately
+             * AUTHOR: Eliseev Dmitry
+             * */
+            public final void stop() {
+                try {
+                    m_Reader.close();
+                } catch (final Exception e) {
+                    LoggingUtility.error("Can't close input stream: " + e.getMessage());
+                }
+            } /* End of 'ClientReceiver::stop' method */
+        } /* End of 'ClientReceiver' class */
+
+        /* Client messages sender class definition */
+        private final class ClientSender implements Runnable {
+            /* Client output stream */
+            private ObjectOutputStream m_Writer = null;
+
+            /* *
+             * METHOD: Client sender class constructor
+             * AUTHOR: Eliseev Dmitry
+             * */
+            public ClientSender() {
+                try {
+                    m_Writer = new ObjectOutputStream(m_Socket.getOutputStream());
+                } catch(final Exception e) {
+                    LoggingUtility.error("Can't open output stream: " + e.getMessage());
+                }
+            } /* End of 'ClientSender::ClientSender' method */
+
+            /* *
+             * METHOD: Blocks running and wait for notification from server
+             * AUTHOR: Eliseev Dmitry
+             * */
+            private void waitForMessages() {
+                synchronized (m_SenderMonitor) {
+                    try {
+                        m_SenderMonitor.wait();
+                    } catch (final InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)